src/Domain/SegmentId.hpp

```diff
@@ -96,8 +104,9 @@ class SegmentId {
   void pup(PUP::er& p) noexcept;  // NOLINT
 
  private:
-  size_t refinement_level_ = std::numeric_limits<size_t>::max();
-  size_t index_ = std::numeric_limits<size_t>::max();
+  unsigned block_id_ : block_id_bits;
```

Comment:
The size of the class `ElementId` class cannot be larger than 3 `int`s, which means for 3d we only get one `int` (32 bits) for each `SegmentId`. Furthermore, Charm++ needs the class to be a POD because it does weird placement new things with it. I'm not exactly sure what you mean with the wold-aligned data and shifting, so if the info I gave here doesn't answer your question then I'll need you to expand a bit please :)

Reply:
I see. I think I wasn't terribly clear -- For the sake of understanding the design choices (I'm not sure if I'd like the following any better) we could supply charm just with an array of `int`s, and create generating functions which transform `block_id_`, `index_`, and `refinement_level_` to and from that array of ints representation. Unfortunately, that translation would probably amount to something similar to manually manipulating the bit representations anyway.

Reply:
For my understanding (this was also a trouble point for me when looking over this PR)... where does the constraint of `ElementId` being no larger than 3 ints come from? Does charm++ only allow us to use a 12-byte "object" to index a charm array with?

Reply:
yes, it is a Charm++ constraint; the index has to be of size 1, 2, or 3 ints

Reply:
see [the Charm documentation](https://charm.readthedocs.io/en/latest/charm++/manual.html#user-defined-array-indices)

Reply:
Note we currently have `ElementId` which is how anyone would want to define this object, and `ElementIndex` which satisfies the requirements of a Charm++ array index.  Both hold the same information, and we cast back and forth when dealing with Charm++.  But we (@nilsdeppe @wthrowe and myself) were finding it annoying to have two separate objects that held the same content...

Reply:
I see - thanks for the context, I think that clears up most of my concerns with this.

Reply:
Thanks @kidder this was helpful!

Reply:
Okay, so the conclusion here is, "Yea, it sucks that we have to deal with bits but this is the hand Charm++ has dealt as and we'd rather have 1 class than 2?" I just want to make sure if there is a change request after this discussion that I know what to do :)

Reply:
Yeah, that sums up my feeling on this. I have no further change request here.

Reply:
I have some documentation thoughts (see my review) but no code changes on this.